torchvision/csrc/io/image/cuda/decode_jpeg_cuda.cpp

```diff
@@ -0,0 +1,185 @@
+#include "decode_jpeg_cuda.h"
+
+#include <ATen/ATen.h>
+
+#if NVJPEG_FOUND
+#include <ATen/cuda/CUDAContext.h>
+#include <c10/cuda/CUDAGuard.h>
+#include <nvjpeg.h>
+#endif
+
+#include <string>
+
+namespace vision {
+namespace image {
+
+#if !NVJPEG_FOUND
+
+torch::Tensor decode_jpeg_cuda(
+    const torch::Tensor& data,
+    ImageReadMode mode,
+    torch::Device device) {
+  TORCH_CHECK(
+      false, "decode_jpeg_cuda: torchvision not compiled with nvJPEG support");
+}
+
+#else
+
+namespace {
+static nvjpegHandle_t nvjpeg_handle = nullptr;
+}
+
+torch::Tensor decode_jpeg_cuda(
+    const torch::Tensor& data,
+    ImageReadMode mode,
+    torch::Device device) {
+  TORCH_CHECK(data.dtype() == torch::kU8, "Expected a torch.uint8 tensor");
+
+  TORCH_CHECK(
+      !data.is_cuda(),
+      "The input tensor must be on CPU when decoding with nvjpeg")
+
+  TORCH_CHECK(
+      data.dim() == 1 && data.numel() > 0,
+      "Expected a non empty 1-dimensional tensor");
+
+  TORCH_CHECK(device.is_cuda(), "Expected a cuda device")
+
+  at::cuda::CUDAGuard device_guard(device);
+
+  // Create global nvJPEG handle
+  std::once_flag nvjpeg_handle_creation_flag;
+  std::call_once(nvjpeg_handle_creation_flag, []() {
+    if (nvjpeg_handle == nullptr) {
+      nvjpegStatus_t create_status = nvjpegCreateSimple(&nvjpeg_handle);
+
+      if (create_status != NVJPEG_STATUS_SUCCESS) {
+        // Reset handle so that one can still call the function again in the
+        // same process if there was a failure
+        free(nvjpeg_handle);
```

Comment:
Since it's an opaque handle I think the use of `free()` may not be correct unless it's documented as being supported.

(I would hope that it simply leaves the handle as null if initialisation fails, although I don't see that in the docs - [here it just reinits the handle without any freeing when hw backend fails though](https://github.com/NVIDIA/CUDALibrarySamples/blob/eb89d592ee4c05a9b358cdc1888e29bb00760d99/nvJPEG/nvJPEG-Decoder/nvjpegDecoder.cpp#L334))

Reply:
Hm that's a good point. What would you recommend instead of `free`?

Before pushing this I did a quick test by inserting

```C++
       nvjpegStatus_t create_status = nvjpegCreateSimple(&nvjpeg_handle);
       create_status = NVJPEG_STATUS_NOT_INITIALIZED;   // <- this
```

and all the tests  were failing gracefully  with `E   RuntimeError: nvjpegCreateSimple failed: 1`. Since I was running the tests with `pytest test/test_image.py -k cuda` they were all in the same process and `pytest` was just catching the `RuntimeErrors`, so I assumed it was OK.